File: src/dns-server.js
import dgram from "node:dgram";
import net from "node:net";
import dnsPacket from "dns-packet";
import { EXPRSN_TLD } from "./models.js";

export const DEFAULT_TTL = 60;

export class DNSServer {
  constructor({
    storage,
    host = "::",
    port = 5353,
    ttl = DEFAULT_TTL,
    nsHost = `ns1.${EXPRSN_TLD}`,
    adminEmail = `hostmaster.${EXPRSN_TLD}`,
    logger = null,
  } = {}) {
    this.storage = storage;
    this.host = host;
    this.port = port;
    this.ttl = ttl;
    this.nsHost = nsHost;
    this.adminEmail = adminEmail;
    this.logger = logger;
    this.socket = null;
    this._serial = 1;
  }

  async start() {
    const type = net.isIPv6(this.host) || this.host === "::" ? "udp6" : "udp4";
    this.socket = dgram.createSocket({ type, reuseAddr: true });
    this.socket.on("message", (msg, rinfo) => this._onMessage(msg, rinfo));
    this.socket.on("error", (err) => {
      this.logger?.error?.(`dns socket error: ${err.message}`);
    });
    await new Promise((resolve, reject) => {
      this.socket.once("error", reject);
      this.socket.bind(this.port, this.host, () => {
        this.socket.removeListener("error", reject);
        resolve();
      });
    });
    const addr = this.socket.address();
    this.port = addr.port;
    return addr;
  }

  stop() {
    return new Promise((resolve) => {
      if (!this.socket) return resolve();
      this.socket.close(() => {
        this.socket = null;
        resolve();
      });
    });
  }

  _soa() {
    return {
      name: EXPRSN_TLD,
      type: "SOA",
      ttl: this.ttl,
      data: {
        mname: this.nsHost,
        rname: this.adminEmail,
        serial: this._serial,
        refresh: 3600,
        retry: 600,
        expire: 86400,
        minimum: this.ttl,
      },
    };
  }

  _onMessage(msg, rinfo) {
    let query;
    try {
      query = dnsPacket.decode(msg);
    } catch (err) {
      this.logger?.warn?.(`dns decode failed: ${err.message}`);
      return;
    }
    const reply = this.resolve(query);
    try {
      const buf = dnsPacket.encode(reply);
      this.socket.send(buf, rinfo.port, rinfo.address);
    } catch (err) {
      this.logger?.error?.(`dns encode/send failed: ${err.message}`);
    }
  }

  resolve(query) {
    const q = query.questions?.[0];
    const reply = {
      id: query.id,
      type: "response",
      flags: dnsPacket.AUTHORITATIVE_ANSWER | dnsPacket.RECURSION_DESIRED,
      questions: q ? [q] : [],
      answers: [],
      authorities: [],
      additionals: [],
    };
    if (!q) {
      reply.flags |= 1; // FORMERR
      return reply;
    }

    const name = (q.name || "").toLowerCase().replace(/\.+$/, "");
    const qtype = q.type;

    if (name !== EXPRSN_TLD && !name.endsWith("." + EXPRSN_TLD)) {
      reply.flags |= 5; // REFUSED
      return reply;
    }

    if (name === EXPRSN_TLD) {
      if (qtype === "SOA" || qtype === "ANY") {
        reply.answers.push(this._soa());
      }
      if (qtype === "NS" || qtype === "ANY") {
        reply.answers.push({
          name: EXPRSN_TLD,
          type: "NS",
          ttl: this.ttl,
          data: this.nsHost,
        });
      }
      if (reply.answers.length === 0) {
        reply.authorities.push(this._soa());
      }
      return reply;
    }

    const record = this.storage.get(name);
    if (!record) {
      reply.flags |= 3; // NXDOMAIN
      reply.authorities.push(this._soa());
      return reply;
    }

    let answered = false;
    if (qtype === "AAAA" || qtype === "ANY") {
      reply.answers.push({
        name: record.token,
        type: "AAAA",
        ttl: this.ttl,
        data: record.ipv6,
      });
      answered = true;
    }
    if (qtype === "NS" || qtype === "ANY") {
      reply.answers.push({
        name: record.token,
        type: "NS",
        ttl: this.ttl,
        data: record.dnsAddress,
      });
      answered = true;
    }
    if (qtype === "TXT" || qtype === "ANY") {
      const parts = [`contact=${record.email}`];
      if (record.certificateId) parts.push(`cert=${record.certificateId}`);
      if (record.ownerId) parts.push(`owner=${record.ownerId}`);
      reply.answers.push({
        name: record.token,
        type: "TXT",
        ttl: this.ttl,
        data: parts.map((s) => Buffer.from(s, "utf8")),
      });
      answered = true;
    }
    if ((qtype === "A" || qtype === "ANY") && net.isIPv4(record.dnsAddress)) {
      reply.answers.push({
        name: record.token,
        type: "A",
        ttl: this.ttl,
        data: record.dnsAddress,
      });
      answered = true;
    }
    if (!answered) {
      reply.authorities.push(this._soa());
    }
    return reply;
  }
}

File: src/exprsndns/dns_server.py
"""Authoritative DNS server for the .exprsn TLD."""

from __future__ import annotations

import ipaddress
import logging
import threading

from dnslib import QTYPE, RR, A, AAAA, NS, RCODE, SOA, TXT, DNSLabel
from dnslib.server import BaseResolver, DNSLogger, DNSServer

from .models import EXPRSN_TLD
from .storage import Storage

log = logging.getLogger(__name__)

DEFAULT_TTL = 60


class ExprsnResolver(BaseResolver):
    """Resolve .exprsn tokens from the backing Storage."""

    def __init__(
        self,
        storage: Storage,
        ns_host: str = f"ns1.{EXPRSN_TLD}",
        admin_email: str = f"hostmaster.{EXPRSN_TLD}",
        ttl: int = DEFAULT_TTL,
    ) -> None:
        self.storage = storage
        self.ns_host = ns_host
        self.admin_email = admin_email
        self.ttl = ttl
        self._serial = 1

    def _soa_record(self, zone: str) -> RR:
        return RR(
            rname=zone,
            rtype=QTYPE.SOA,
            ttl=self.ttl,
            rdata=SOA(
                mname=self.ns_host,
                rname=self.admin_email,
                times=(self._serial, 3600, 600, 86400, self.ttl),
            ),
        )

    def resolve(self, request, handler):  # type: ignore[override]
        reply = request.reply()
        qname = request.q.qname
        qtype = QTYPE[request.q.qtype]
        name = str(qname).rstrip(".").lower()

        if not (name == EXPRSN_TLD or name.endswith("." + EXPRSN_TLD)):
            reply.header.rcode = RCODE.REFUSED
            return reply

        zone = DNSLabel(EXPRSN_TLD)

        # Apex queries for the zone itself.
        if name == EXPRSN_TLD:
            if qtype in ("SOA", "ANY"):
                reply.add_answer(self._soa_record(zone))
            if qtype in ("NS", "ANY"):
                reply.add_answer(
                    RR(zone, QTYPE.NS, ttl=self.ttl, rdata=NS(self.ns_host))
                )
            if not reply.rr:
                reply.add_auth(self._soa_record(zone))
            return reply

        record = self.storage.get(name)
        if record is None:
            reply.header.rcode = RCODE.NXDOMAIN
            reply.add_auth(self._soa_record(zone))
            return reply

        answered = False
        if qtype in ("AAAA", "ANY"):
            reply.add_answer(
                RR(qname, QTYPE.AAAA, ttl=self.ttl, rdata=AAAA(record.ipv6))
            )
            answered = True
        if qtype in ("NS", "ANY"):
            reply.add_answer(
                RR(qname, QTYPE.NS, ttl=self.ttl, rdata=NS(record.dns_address))
            )
            answered = True
        if qtype in ("TXT", "ANY"):
            reply.add_answer(
                RR(
                    qname,
                    QTYPE.TXT,
                    ttl=self.ttl,
                    rdata=TXT(f"contact={record.email}".encode("ascii")),
                )
            )
            answered = True
        if qtype in ("A", "ANY"):
            # If the stored dns_address is an IPv4 literal, surface it.
            try:
                ipv4 = ipaddress.IPv4Address(record.dns_address)
                reply.add_answer(
                    RR(qname, QTYPE.A, ttl=self.ttl, rdata=A(str(ipv4)))
                )
                answered = True
            except ValueError:
                pass

        if not answered:
            reply.add_auth(self._soa_record(zone))
        return reply


class ExprsnDNSServer:
    """Wrapper that runs a UDP DNS server bound to the given address."""

    def __init__(
        self,
        storage: Storage,
        host: str = "::",
        port: int = 5353,
        ttl: int = DEFAULT_TTL,
        quiet: bool = True,
    ) -> None:
        self.resolver = ExprsnResolver(storage, ttl=ttl)
        logger = DNSLogger(prefix=False) if not quiet else DNSLogger("-request,-reply,-truncated,-error", False)
        self._server = DNSServer(
            self.resolver,
            address=host,
            port=port,
            logger=logger,
        )
        self._thread: threading.Thread | None = None

    def start(self) -> None:
        self._server.start_thread()
        self._thread = self._server.thread

    def stop(self) -> None:
        self._server.stop()

    def serve_forever(self) -> None:
        self._server.start()
